src/System.IO.Pipes/src/System/IO/Pipes/NamedPipeServerStream.Windows.cs

```diff
@@ -53,16 +69,28 @@ public sealed partial class NamedPipeServerStream : PipeStream
                 maxNumberOfServerInstances = 255;
             }
 
-            Interop.Kernel32.SECURITY_ATTRIBUTES secAttrs = PipeStream.GetSecAttrs(inheritability);
-            SafePipeHandle handle = Interop.Kernel32.CreateNamedPipe(fullPipeName, openMode, pipeModes,
-                maxNumberOfServerInstances, outBufferSize, inBufferSize, 0, ref secAttrs);
+            var pinningHandle = new GCHandle();
+            Interop.Kernel32.SECURITY_ATTRIBUTES secAttrs = PipeStream.GetSecAttrs(inheritability, pipeSecurity, ref pinningHandle);
 
-            if (handle.IsInvalid)
+            try
             {
-                throw Win32Marshal.GetExceptionForLastWin32Error();
-            }
+                SafePipeHandle handle = Interop.Kernel32.CreateNamedPipe(fullPipeName, openMode, pipeModes,
+                    maxNumberOfServerInstances, outBufferSize, inBufferSize, 0, ref secAttrs);
 
-            InitializeHandle(handle, false, (options & PipeOptions.Asynchronous) != 0);
+                if (handle.IsInvalid)
+                {
+                    throw Win32Marshal.GetExceptionForLastWin32Error();
+                }
+
+                InitializeHandle(handle, false, (options & PipeOptions.Asynchronous) != 0);
+            }
+            finally
+            {
+                if (pinningHandle.IsAllocated)
```

Comment:
Docs say `Use this property when using Weak handles to determine if the GCHandle is still available.` . In this case you are using `GCHandleType.Pinned` not Weak. Is it needed?

Also shouldn't `Interop.Kernel32.SECURITY_ATTRIBUTES secAttrs = PipeStream.GetSecAttrs(inheritability, pipeSecurity, ref pinningHandle);` be inside the try, in case it throws ?

Reply:
The pinningHandle.IsAvailable was per @JeremyKuhne suggestion. Should we then remove this if per the documentantion?

And yes I’ll move the GetSecAttrs call to the try statement.